feat(pricing-table): create pricing table component

jkaeser opened the pull request:
### Related Ticket(s)

Resolves https://github.com/carbon-design-system/carbon-for-ibm-dotcom/issues/8393

### Description

Adds Pricing Table web component.

### Changelog

**New**

- Adds `DDSPricingTable` and supporting web components.
- Creates Storybook stories and documentation for Pricing Table.

**Changed**

- Adds class names to selectors in `DDSStructuredList`'s stylesheet.
- Modifies `structured-list-colspan` mixin to accommodate `--default-cols` CSS custom property.
- Adds overflow behavior to Structured List (which is inherited by Pricing Table) to accommodate wide tables.

<!-- React and Web Component deploy previews are enabled by default. -->
<!-- To enable additional available deploy previews, apply the following -->
<!-- labels for the corresponding package: -->
<!-- *** "test: e2e": Codesandbox examples and e2e integration tests -->
<!-- *** "package: services": Services -->
<!-- *** "package: utilities": Utilities -->
<!-- *** "RTL": React / Web Components (RTL) -->
<!-- *** "feature flag": React / Web Components (experimental) -->

jkaeser commented on packages/web-components/src/components/pricing-table/pricing-table.scss:
```diff
@@ -0,0 +1,11 @@
+// LICENSE file in the root directory of this source tree.
+//
+
+// @import '@carbon/ibmdotcom-styles/scss/components/pricing-table/pricing-table';
+@import '../../../../styles/scss/components/pricing-table/pricing-table';
```
Temporary while in active development.

jkaeser commented:
While this is still a WIP, I understand we want to get #8393 in this week if at all possible, and I think things are in a good enough place to do some preliminary design QA if you have the time, @oliviaflory. Specifically, I could use guidance for:

1. **Column widths:** I noticed in the design that columns' widths change depending on how many there are in the table. Could you map out how these should be sized?
2. **Tag link and tooltip toggle interactions:** Each of these can be interacted with, but I didn't find specs for their interaction states (e.g. on hover). Do you have any thoughts on these?

Some of the more glaring style issues I'm still working on:
- Mobile styling issues
- Group header widths at 6 columns
- lining up tags, whether or not the markup is present

jkaeser commented:
@oliviaflory @jwitkowski79 One more question! The design token specified for the table highlight label text, `text-01`, does not work for the dark themes. Is there another token I can use instead?

<img width="1332" alt="Screen Shot 2022-04-05 at 4 23 42 PM" src="https://user-images.githubusercontent.com/12532727/161842872-522df381-14ea-44f8-900d-7d4b3410f441.png">

oliviaflory commented:
> @oliviaflory @jwitkowski79 One more question! The design token specified for the table highlight label text, text-01, does not work for the dark themes. Is there another token I can use instead?

Good catch @jkaeser the text-01 shouldn't work on the light theme either so I'm surprised we didn't see that there too.
light themes #text-01 = #161616 which wouldn't work on top of the dark highlight.

Tokens should not be changing between themes, all components should be speced with a single theme, so I'm curious what we have that text set to in the white theme that's allowing it to work?
<img width="372" alt="Screen Shot 2022-04-05 at 5 28 03 PM" src="https://user-images.githubusercontent.com/45692622/161852725-ae59390d-e5bb-4dbd-ba14-905cc3f94d92.png">


We should update the specs to follow a single theme:
background: `$ui-05`
text: `$inverse-01`

andy-blum commented on packages/web-components/src/components/pricing-table/pricing-table.ts:
```diff
@@ -0,0 +1,98 @@
+      })
+      .catch(() => {
+        wrapper.style.marginTop = '';
+      });
```
I don't think any part of this is asynchronous - should this just be if/else logic?

andy-blum commented on packages/web-components/src/components/structured-list/structured-list.scss:
```diff
@@ -1,5 +1,5 @@
 //
-// Copyright IBM Corp. 2020, 2021
+// Copyright IBM Corp. 2020, 2022
```
thanks 😬

andy-blum commented on packages/styles/scss/components/pricing-table/_pricing-table.scss:
```diff
@@ -0,0 +1,225 @@
+      width: 100%;
+      height: 1px;
+      background-color: $ui-03;
+      box-shadow: 50vw 0 $ui-03, 100vw 0 $ui-03, 150vw 0 $ui-03, 200vw 0 $ui-03,
```
Did you not find a way to get the sass function working?

jkaeser replied:
I couldn't figure it out quickly, so I shifted priority to getting this PR into a reviewable state. I'll look into it more today.

andy-blum commented on packages/web-components/src/components/pricing-table/__stories__/pricing-table.stories.ts:
```diff
@@ -0,0 +1,281 @@
+  `;
+};
+
+export const WithSubheaders = ({ parameters }) => {
```
We need to title-case these variations in storyboo
<img width="224" alt="Screen Shot 2022-04-05 at 5 46 11 PM" src="https://user-images.githubusercontent.com/25532785/161855228-6a4cc6c7-24a6-4c7c-b84a-f6a3cd9f56eb.png">
k

andy-blum reviewed (changes requested):
Seems like a great piece of work here, though I didn't stress test it too much. Thanks for backporting some of those style issues to the structured-list component. A few changes/questions below

proeung commented on packages/web-components/src/components/pricing-table/__stories__/pricing-table.stories.ts:
```diff
@@ -0,0 +1,287 @@
+            <bx-list-item>consectetur retention adispiscing elit sed do eiusm Eiusmod tempor</bx-list-item>
+          </bx-unordered-list>
+        </dds-pricing-table-header-cell-description>
+        <dds-pricing-table-header-cell-cta href="https://www.carbondesignsystem.com/all-about-carbon/what-is-carbon/"
```
I know that @jwitkowski79's design spec doesn't include the CTA icon, however, since this component is extending from `DDSButtonCTA` and having seen the node structure of the CTA button in AEM, we'll need to add `cta-type` which will render different SVGs depending on the selection (see - https://carbon-design-system.github.io/carbon-for-ibm-dotcom/canary/web-components/?path=/story/components-cta--button). I think in this storybook preview, we can just add `cta-type="local"`.

jkaeser replied:
Thank you for clarifying! I added that attribute to `<dds-pricing-table-header-cell-cta>` elements in the documentation and each of the examples.

proeung reviewed (changes requested):
@jkaeser Good work putting this component together! I see that there was a lot of work and attention to detail that has been put into this PR! I just have a couple of comments, but overall this is looking solid.

---------------------

@dratzenboeck-ibm @EAVilleda Can you help review this PR to ensure that we're not missing any dependencies (eg. slot styling, etc) needed on the AEM side?

jkaeser commented on packages/web-components/examples/codesandbox/components/pricing-table/cdn.html:
```diff
@@ -0,0 +1,38 @@
+<div class="bx--grid">
+  <div class="bx--row">
+    <div class="bx--col-sm-4 bx--col-lg-12">
+      <dds-pricing-table></dds-pricing-table>
```
Done!

dratzenboeck-ibm commented:
> @dratzenboeck-ibm @EAVilleda Can you help review this PR to ensure that we're not missing any dependencies (eg. slot styling, etc) needed on the AEM side?

I can't really tell what is needed in detail before starting with the AEM implementation. Regarding slots it would be generally helpful to apply css via the slot name instead of the tag name, so that we are able to wrap the tags with vanilla tags in the author view, as AEM doesn't apply its overlay to custom tags.

jwitkowski79 commented:
@jkaeser Hi! This is looking great! Thanks for all of your hard work on this! Just did a round of design QA. I didn't get a chance yet to test on tablet and mobile but wanted to post what I have so far. I'm on vacation until Monday so hoping @oliviaflory  can step in and help finish the design QA on this component. 

1. Can we switch the info icon to the down chevron? My apologies. This never was updated in the design specs. It should be updated now in [box](https://ibm.ent.box.com/folder/157069618862?s=5qv9zo7jdm4pd7hc7lj5ogqy0czj50ie)
![Components___Pricing_Table_-_Default_⋅_Storybook 2](https://user-images.githubusercontent.com/2040791/162028923-0eae38e1-aece-4d84-959f-a44ac63a269e.png)

2. Looks like we need to increase the bottom padding of cell data to 16px (see attached screenshot)
![Components___Pricing_Table_-_With_subheaders_⋅_Storybook_🔊](https://user-images.githubusercontent.com/2040791/162029039-032b85b5-7a70-42b1-aa88-6db429ca05fe.png)

3. Add 8px padding between text block and info icon
![Components___Pricing_Table_-_Without_Row_Headers_⋅_Storybook (2)](https://user-images.githubusercontent.com/2040791/162029103-c4459668-ee31-4b4f-adec-215363b55eab.png)

4. I know this is a work in progress but is a sticky header still being implemented on scroll down? See [functional and design specs](https://ibm.ent.box.com/folder/157069618862?s=5qv9zo7jdm4pd7hc7lj5ogqy0czj50ie)

5. Is the variation of the Pricing table with content block still being worked on? Seems like it's missing as an option
![component--structured lists-complex--block--specs--max](https://user-images.githubusercontent.com/2040791/162029595-2afbb5f6-02eb-4e2f-82e7-348ecfe71eed.png)

6. In the CTA, can we turn off the arrow icon by default? We wouldn’t recommend using it in our documentation and don’t want to have it on by default. Also, the arrows are behaving oddly. See screenshot of Safari and Chrome
![Components___Pricing_Table_-_Default_⋅_Storybook (1) 2](https://user-images.githubusercontent.com/2040791/162030238-9149f23f-d146-4df2-a48f-c6ab2b52da34.png)
![Components___Pricing_Table_-_Default_⋅_Storybook 3](https://user-images.githubusercontent.com/2040791/162030240-fe6d0ee4-8287-44e0-abbf-d785ed5c7b2a.png)

7. @oliviaflory would love your input on the below grid layout suggestions from @jkaeser to implement
    - 6-8 column table: columns should expand 2 columns
    - 4-5 column table: columns should expand 3 columns
    - 1-3 column table: columns should expand 4 columns

proeung commented:
@jkaeser Per my conversation (see [Slack thread](https://ibm-marketing-comms.slack.com/archives/G01LKRP5RB6/p1649266357514599?thread_ts=1649263741.454539&cid=G01LKRP5RB6)) with @ljcarot and @kennylam, we're going to mark this component behind a "Feature Flag" for now.

Can you add a commit that adds a new boolean and wraps the storybook component via a feature flag (see - https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/main/packages/web-components/docs/feature-flags.md)? Also, let's call it `DDS_PRICING_TABLE`. See [Cloud Masthead](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/main/packages/web-components/src/components/masthead/__stories__/cloud-masthead.stories.ts#L48) as an example.

RichKummer commented:
Took a look at `md` and `sm`

1. Jen mentioned in #6 that the button is behaving oddly. Just wanted to surface this in `md` when set to 2-columns: 
<img width="670" alt="Screen Shot 2022-04-06 at 3 11 45 PM" src="https://user-images.githubusercontent.com/15643582/162051584-2dd588f0-dfd2-4b94-94a9-e7facb37c8ab.png">

2. In the Table heading, the last item in the unordered list has additional padding at the bottom, but this should be removed so that there is 24px padding between the last item and the CTA.
<img width="671" alt="Screen Shot 2022-04-06 at 3 20 14 PM" src="https://user-images.githubusercontent.com/15643582/162052618-8bd0cf60-2712-43a2-8587-f78146d6d125.png">

3. If the tag in the Table heading wraps to a new line, the copy right before the unordered list should maintain a 16px padding between:
<img width="667" alt="Screen Shot 2022-04-06 at 3 32 16 PM" src="https://user-images.githubusercontent.com/15643582/162054776-3ff708b1-841c-49aa-956b-5e682b6f594d.png">

4. When I tried to expand the row for more information on iPad in Safari, the text will appear instantly instead of fading in from the top like on Desktop. When I condense the row, the added content will fade up as expected. This issue only happens when the row is already at the appropriate height - if the row needs to expand the height, the animation will work as intended.

https://user-images.githubusercontent.com/15643582/162058675-4a552dbd-0319-4f95-83ad-e2ca5d0797d0.MP4

RichKummer commented:
Thanks for this @jkaeser ! Took a look at `md` and `sm`:

1. Jen mentioned in #6 that the button is behaving oddly. Just wanted to surface this in `md` when set to 2-columns: 
<img width="670" alt="Screen Shot 2022-04-06 at 3 11 45 PM" src="https://user-images.githubusercontent.com/15643582/162051584-2dd588f0-dfd2-4b94-94a9-e7facb37c8ab.png">

2. In the Table heading, the last item in the unordered list has additional padding at the bottom, but this should be removed so that there is 24px padding between the last item and the CTA.
<img width="671" alt="Screen Shot 2022-04-06 at 3 20 14 PM" src="https://user-images.githubusercontent.com/15643582/162052618-8bd0cf60-2712-43a2-8587-f78146d6d125.png">

3. If the tag in the Table heading wraps to a new line, the copy right before the unordered list should maintain a 16px padding between. Is that right @jwitkowski79 ?
<img width="667" alt="Screen Shot 2022-04-06 at 3 32 16 PM" src="https://user-images.githubusercontent.com/15643582/162054776-3ff708b1-841c-49aa-956b-5e682b6f594d.png">

4. When I tried to expand the row for more information on iPad in Safari, the text will appear instantly instead of fading in from the top like on Desktop. When I condense the row, the added content will fade up as expected. This issue only happens when the row is already at the appropriate height - if the row needs to expand the height, the animation will work as intended.

https://user-images.githubusercontent.com/15643582/162058675-4a552dbd-0319-4f95-83ad-e2ca5d0797d0.MP4


5. When a row is expandable, but the expanded content _does not_ contain a checkmark icon, it looks like there is some additional padding at the top of the row text that should not be there. I didn't see an example in the specs, so wanted to double check that there should not be that added 16px spacing in this instance @jwitkowski79 

<img width="666" alt="Screen Shot 2022-04-06 at 3 55 19 PM" src="https://user-images.githubusercontent.com/15643582/162059989-0ba6fc15-0256-42c8-825a-b853d33f7bee.png">

jkaeser commented:
Thanks for the review, Jen, and sorry to bother on your vacation!

@oliviaflory @RichKummer I've addressed Jen's feedback except for the following:

4. Sticky headers are coming in a followup task. See https://github.com/carbon-design-system/carbon-for-ibm-dotcom/issues/8420.
5. Is this variation something that needs to be demonstrated in Storybook? As I understand it, this represents a composition of two components rather than a feature of the Pricing Table. cc @proeung
6. See Putra's comment: https://github.com/carbon-design-system/carbon-for-ibm-dotcom/pull/8674#discussion_r844000133. @proeung, does the AEM team _need_ this functionality? Currently, the Pricing Table CTAs will respect the `cta-type` attribute, but it sounds like maybe we should disable that in this component to adhere to design specs.

@RichKummer Thank you as well for your review! I've started working through your feedback, and so far I fixed the tag spacing (`#3`).

proeung reviewed (approved):
LGTM! Thanks for addressing my feedback!

----------

@kennylam @IgnacioBecerra Can one of you take a look at this feature flag PR? I'd like to get this merge down before the code freeze period. Thanks!

kennylam reviewed (approved):
LGTM thanks @jkaeser!

IgnacioBecerra reviewed (approved):
LGTM, the component looks pretty good, thanks @jkaeser!

kennylam commented:
@jkaeser Looks like this needs to get resolved before this check can pass https://github.com/carbon-design-system/carbon-for-ibm-dotcom/runs/5875018446?check_suite_focus=true#step:6:5905

RichKummer commented:
@jkaeser still seeing a few design issues:

1. The spacing at the bottom of the row should be 16px per Jen's comment - I'm seeing the bottom as 24px
<img width="651" alt="Screen Shot 2022-04-08 at 4 29 07 PM" src="https://user-images.githubusercontent.com/15643582/162525745-b1f0e51e-658f-429e-b78d-3faa532455ba.png">

2. At the `md` breakpoint exactly (672px), the description on the second column moves up to align with the tag. Once I move the browser 1px larger, the text moves to the correct spot - aligning with the descriptive text in the first column.

**672px:**
<img width="602" alt="Screen Shot 2022-04-08 at 4 31 31 PM" src="https://user-images.githubusercontent.com/15643582/162526188-b6ef784c-a9a4-4d4f-a0df-182334a41b55.png">

**673px:**
<img width="571" alt="Screen Shot 2022-04-08 at 4 34 03 PM" src="https://user-images.githubusercontent.com/15643582/162526358-7b3edbf9-a417-415f-9a7d-0bfdf944f4f5.png">

3. The last item in the unordered list in the table header still has 4px bottom spacing - to match the specs, we'd need to remove that spacing so that there is 24px space between the last item and the button
<img width="768" alt="Screen Shot 2022-04-08 at 4 36 49 PM" src="https://user-images.githubusercontent.com/15643582/162526659-e533e880-21c7-4ae9-95c2-a09060b30fef.png">

jkaeser commented:
Good catches @RichKummer! I just pushed some commits that should fix each of those issues with the exception of the list item spacing. That is baked into the `<bx-list-item>` element from Carbon Web Components, and as such I am unable to override its style. I _could_ adjust the margin above the CTA buttons to be 8px less, but that would cause spacing issues when the last element in the text area isn't a `<bx-list-item>` and doesn't set exactly 8px of bottom margin. How do you prefer I proceed?

RichKummer commented:
> Good catches @RichKummer! I just pushed some commits that should fix each of those issues with the exception of the list item spacing. That is baked into the `<bx-list-item>` element from Carbon Web Components, and as such I am unable to override its style. I _could_ adjust the margin above the CTA buttons to be 8px less, but that would cause spacing issues when the last element in the text area isn't a `<bx-list-item>` and doesn't set exactly 8px of bottom margin. How do you prefer I proceed?

I'll write a separate issue for this one, I'd rather not make an update here that causes other issues. Thanks @jkaeser, I'll take a look when the build finishes!
